homeassistant/components/fritz/common.py

```diff
@@ -0,0 +1,230 @@
+"""Support for AVM FRITZ!Box classes."""
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import logging
+from typing import Any, Dict, Optional
+
+# pylint: disable=import-error
+from fritzconnection import FritzConnection
+from fritzconnection.lib.fritzhosts import FritzHosts
+from fritzconnection.lib.fritzstatus import FritzStatus
+
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    DEFAULT_HOST,
+    DEFAULT_PORT,
+    DEFAULT_USERNAME,
+    DOMAIN,
+    TRACKER_SCAN_INTERVAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class Device:
+    """FRITZ!Box device class."""
+
+    mac: str
+    ip_address: str
+    name: str
+
+
+class FritzBoxTools:
+    """FrtizBoxTools class."""
+
+    def __init__(
+        self,
+        hass,
+        password,
+        username=DEFAULT_USERNAME,
+        host=DEFAULT_HOST,
+        port=DEFAULT_PORT,
+    ):
+        """Initialize FritzboxTools class."""
+        self._cancel_scan = None
+        self._device_info = None
+        self._devices: Dict[str, Any] = {}
+        self._unique_id = None
+        self.connection = None
+        self.error = None
+        self.fritzhosts = None
+        self.fritzstatus = None
+        self.hass = hass
+        self.host = host
+        self.password = password
+        self.port = port
+        self.username = username
+
+    async def async_setup(self):
+        """Wrap up FritzboxTools class setup."""
+        return await self.hass.async_add_executor_job(self.setup)
+
+    def setup(self):
+        """Set up FritzboxTools class."""
+
+        self.connection = FritzConnection(
+            address=self.host,
+            port=self.port,
+            user=self.username,
+            password=self.password,
+            timeout=60.0,
+        )
+
+        self.fritzstatus = FritzStatus(fc=self.connection)
+        if self._unique_id is None:
+            self._unique_id = self.connection.call_action("DeviceInfo:1", "GetInfo")[
+                "NewSerialNumber"
+            ]
+
+        self._device_info = self._fetch_device_info()
+
+    def start(self):
+        """Start FritzHosts connection."""
+        self.fritzhosts = FritzHosts(fc=self.connection)
+
+        self.scan_devices()
+
+        self._cancel_scan = async_track_time_interval(
```

Comment:
```suggestion
        self._cancel_scan = track_time_interval(
```

This need to be the non-async version since its happening in the executor

Reply:
or you could make an `async_start` and do this in there and wrap the `scan_devices` and fritzhosts creation in an executor call that way you can make `unload` async safe to call in __init__